### PR TITLE
1731 Added endpoint for getting owned connections

### PR DESCRIPTION
**Pull request by tnickelsen**

### CODE REVIEW

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/ConnectionsController.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+using API.Extensions;
+using API.Models;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+
+namespace API.Controllers;
+
+[Authorize]
+[ApiController]
+[Route("api/connections")]
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/Controllers/ConnectionsController.cs:14]: This requires that a ingress-route is defined in eo-base-environment

File: domains/transfer-agreements/TransferAgreements.API/API/Data/ApplicationDbContext.cs [hunk 2]
```diff
@@ -12,4 +12,5 @@ public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options) : ba
 
     public DbSet<TransferAgreement> TransferAgreements { get; set; }
     public DbSet<TransferAgreementHistoryEntry> TransferAgreementHistoryEntries { get; set; }
+    public DbSet<Connection> Connections { get; set; }
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/Data/ApplicationDbContext.cs:15]: I strongly suggest that you make the necessary C4 container diagrams and maybe component diagrams for the Transfer agreement domain as part of this PR.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs [hunk 3]
```diff
@@ -135,9 +148,27 @@ private static async Task TruncateTables(ApplicationDbContext dbContext)
...
+        var agreementQuery = $"INSERT INTO \"{connectionTable}\" (\"Id\", \"CompanyId\", \"CompanyTin\", \"OwnerId\") VALUES (@Id, @CompanyId, @CompanyTin, @OwnerId)";
+        var agreementFields = new[]
+        {
+            new NpgsqlParameter("Id", connection.Id),
+            new NpgsqlParameter("CompanyId", connection.CompanyId),
+            new NpgsqlParameter("CompanyTin", connection.CompanyTin),
+            new NpgsqlParameter("OwnerId", connection.OwnerId)
+        };
+
+        await dbContext.Database.ExecuteSqlRawAsync(agreementQuery, agreementFields);
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs:171]: Why is this inserted using Raw SQL?

File: domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+using System;
+
+namespace API.Models;
+
+public class Connection
+{
+    public Guid Id { get; set; }
+    public Guid CompanyId { get; set; }
+    public string CompanyTin { get; set; }
+    public Guid OwnerId { get; set; }
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs:10]: What is Owner here? I am not I understand how a connection between two companies can have a owner.

File: domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+using System;
+
+namespace API.Models;
+
+public class Connection
+{
+    public Guid Id { get; set; }
+    public Guid CompanyId { get; set; }
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs:8]: Is the name CompanyId the best name for this property? If I am right, then this is to be populated with the SubjectId in the JWT-token - which means that it is the Id assigned by the Auth-service. So I am fearing that AuthService may call this, e.g. OrganizationId. Then we have 3 different terms for the same thing.

File: domains/transfer-agreements/TransferAgreements.API/API/ApiModels/Responses/ConnectionsResponse.cs [hunk 6]
```diff
@@ -0,0 +1,5 @@
+using System.Collections.Generic;
+
+namespace API.ApiModels.Responses;
+
+public record ConnectionsResponse(List<ConnectionDto> Connections);
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/ApiModels/Responses/ConnectionsResponse.cs:5]: I suggest that `Connections` is renamed to `Result` so it is the same as for the TransferAgreementResponse.
We do not have an API guideline determining the structure for returning a list of entities. But let's try to be consistent with what is currently there.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs [hunk 7]
```diff
@@ -81,6 +82,18 @@ public async Task SeedData(IEnumerable<TransferAgreement> transferAgreements)
         }
     }
 
+    public async Task SeedConnections(IEnumerable<Connection> connections)
+    {
+        using var scope = Services.CreateScope();
+        var dbContext = scope.ServiceProvider.GetRequiredService<ApplicationDbContext>();
+        await TruncateTables(dbContext);
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs:89]: It is a bit funky for me, that when seeding connections, then tables are truncated for Transfer Agreements.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs [hunk 8]
```diff
@@ -81,6 +82,18 @@ public async Task SeedData(IEnumerable<TransferAgreement> transferAgreements)
         }
     }
 
+    public async Task SeedConnections(IEnumerable<Connection> connections)
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs:85]: The same method for TransferAgreements is named `SeedData`. There should probably be some renaming of `SeedData`

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs [hunk 9]
```diff
@@ -18,6 +18,7 @@
 using Microsoft.Extensions.Hosting;
 using Microsoft.IdentityModel.Tokens;
 using Npgsql;
+using Org.BouncyCastle.Crypto.Agreement;
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs:21]: Why is this added?

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionsControllerTests.cs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        });
+
+        var client = factory.CreateAuthenticatedClient(sub: sub.ToString());
+
+        var get = await client.GetAsync($"api/connections");
+        get.EnsureSuccessStatusCode();
+        var response = JsonConvert.DeserializeObject<ConnectionsResponse>(await get.Content.ReadAsStringAsync());
+
+        response.Should().NotBeNull();
+        response.Connections.Should().HaveCount(1);
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionsControllerTests.cs:54]: There is no assertion on the returned connections is the "owned" connection and not the "not owned" connections.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs (hunk 3 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs:171]: Excellent question. Because that's what was already used - I've corrected it

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs (hunk 9 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs:21]: Not sure - removed

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs (hunk 8 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs:85]: Renamed

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs (hunk 7 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs:89]: Made separate truncate for connections only

File: domains/transfer-agreements/TransferAgreements.API/API/ApiModels/Responses/ConnectionsResponse.cs (hunk 6 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API/ApiModels/Responses/ConnectionsResponse.cs:5]: Agreed - renamed

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionsControllerTests.cs (hunk 10 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionsControllerTests.cs:54]: Added

File: domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs (hunk 4 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs:10]: Owner is the 'accepter' subject id. When two connections are made when the invite is accepted - one for the receiver and one for the sender

File: domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs (hunk 5 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs:8]: CompanyId is then the subject id of the one who invited

Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs:8]: We discussed it - Auth is using company as well

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/ConnectionsController.cs (hunk 1 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API/Controllers/ConnectionsController.cs:14]: I did not know that. Will you show me at some point?

File: domains/transfer-agreements/TransferAgreements.API/API/Data/ApplicationDbContext.cs (hunk 2 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API/Data/ApplicationDbContext.cs:15]: Sure - where do you suggest we put it?

Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/Data/ApplicationDbContext.cs:15]: Have you considered using database schemas to mark the difference between Connections and TA?

File: domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs (hunk 4 above)
Comment (duizer) [reply domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs:10]: I am reacting to the concept of "owning" a connection. I do not understand why there is an ownership on a connection...?

Furthermore, when you explain why, you are not listing a ownership relationship, but only that it is a connection between two companies and one is `sending` and the other is `receiving`. Why not use those terms `sending`/`receiving`? Or `inviter`/`invitee`?

File: domains/transfer-agreements/TransferAgreements.API/API/Data/ApplicationDbContext.cs (hunk 2 above)
Comment (duizer) [reply domains/transfer-agreements/TransferAgreements.API/API/Data/ApplicationDbContext.cs:15]: Christoffer did some work on the C4 diagrams, so ask him for a status. Otherwise, there is doc\architecture\domains\transfer-agreements.md in energy-origin.

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/ConnectionsController.cs (hunk 1 above)
Comment (duizer) [reply domains/transfer-agreements/TransferAgreements.API/API/Controllers/ConnectionsController.cs:14]: Yes. Short answer is add to this file k8s\energy-origin-apps\ingress-routes\base\ingress-routes.yaml in eo-base-environment

File: domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs (hunk 5 above)
Comment (duizer) [reply domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs:8]: ok

File: domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs (hunk 4 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs:10]: Well because with this page I'm viewing _my_ connections. But maybe I'm looking at it wrong - how do you propose to do it?

Because from a connections standpoint it does not make sense to know who the invite and invitee is

File: domains/transfer-agreements/TransferAgreements.API/API/Data/ApplicationDbContext.cs (hunk 2 above)
Comment (tnickelsen) [reply domains/transfer-agreements/TransferAgreements.API/API/Data/ApplicationDbContext.cs:15]: No - I'm unsure what that would add?

File: doc/diagrams/c4-model/model.dsl [hunk 11]
```diff
@@ -51,5 +51,25 @@ measurementsDomain = group "Measurements Domain" {
 }
 
 transferAgreementsDomain = group "Transfer Agreements Domain" {
+    taApi = container "Transfer Agreement API" "" ".NET Web Api" {
+        connectionsController = component "Connections Controller" "Allows users to see connections of their company." ".NET Web Api Controller"
+        connectionInvitesController = component "Connection-Invites Controller" "Allows users to create connections with other companies" ".NET Web Api Controller"
+        TransferAgreementsController = component "Transfer Agreements Controller" "Allows users to create transfer agreements with other companies" ".NET Web Api Controller"
```
Comment (duizer) [inline doc/diagrams/c4-model/model.dsl:57]: Should we call these "Controllers". Will people confuse it with a Controller-class?

Comment (duizer) [inline doc/diagrams/c4-model/model.dsl:57]: This also communicates with PO Wallet

File: doc/diagrams/c4-model/model.dsl [hunk 12]
```diff
@@ -51,5 +51,25 @@ measurementsDomain = group "Measurements Domain" {
...
+        TransferAgreementAutomation = component "Transfer Agreements Automation" "Transfers certificates within a given transfer agreement" ".NET BackgroundService" {
+            this -> po "Transfers certificates"
```
Comment (duizer) [inline doc/diagrams/c4-model/model.dsl:59]: There is no system called "Project Origin" - it should be a system for Registry and one for Wallet System

File: doc/diagrams/c4-model/model.dsl [hunk 13]
```diff
@@ -51,5 +51,25 @@ measurementsDomain = group "Measurements Domain" {
 }
 
 transferAgreementsDomain = group "Transfer Agreements Domain" {
+    taApi = container "Transfer Agreement API" "" ".NET Web Api" {
```
Comment (duizer) [inline doc/diagrams/c4-model/model.dsl:54]: There is a component missing for the background job that removes old invitations

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/SwaggerTests.GetSwaggerDoc_AppStarted_NoChangesAccordingToSnapshot.verified.txt [hunk 14]
```diff
@@ -344,6 +376,35 @@
...
+            "type": "string",
+            "format": "uuid"
+          },
+          "companyId": {
+            "type": "string",
+            "format": "uuid"
+          },
+          "comnpanyTin": {
+            "type": "string"
+          }
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/SwaggerTests.GetSwaggerDoc_AppStarted_NoChangesAccordingToSnapshot.verified.txt:392]: There is no description/summary for any of these properties. That can be done by commenting on the properties of `ConnectionDto`. Maybe you have to change `ConnectionDto` as I do not think it works for a record with positional syntax

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/ConnectionInvitationsController.cs [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        var companySenderId = Guid.Parse(User.FindSubjectGuidClaim());
+        var companySenderTin = User.FindSubjectTinClaim();
+
+        var newInvitation = new ConnectionInvitation
+        {
+            SenderCompanyId = companySenderId,
+            SenderCompanyTin = companySenderTin,
+        };
+
+        var addedInvitation = await connectionInvitationRepository.AddConnectionInvitation(newInvitation);
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/Controllers/ConnectionInvitationsController.cs:34]: I think it is strange that and "Add"-method returns the same object. Speaking in terms of CQS, then this is a command, but it has the signature of a query.

File: domains/transfer-agreements/TransferAgreements.API/API/Data/IConnectionRepository.cs [hunk 16]
```diff
@@ -0,0 +1,11 @@
+using System;
+using System.Collections.Generic;
+using API.Models;
+
+namespace API.Data
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/Data/IConnectionRepository.cs:5]: Please use file-scoped namespaces and not block-scoped

File: domains/transfer-agreements/TransferAgreements.API/API/Data/IConnectionRepository.cs [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+{
+    public interface IConnectionRepository
+    {
+        List<Connection> GetCompanyConnections(Guid ownerId);
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/Data/IConnectionRepository.cs:9]: Is there a reason for why this is not async?

File: domains/transfer-agreements/TransferAgreements.API/API/Migrations/ApplicationDbContextModelSnapshot.cs [hunk 18]
```diff
@@ -1,148 +1,177 @@
-﻿// <auto-generated />
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/Migrations/ApplicationDbContextModelSnapshot.cs:1]: Since all lines are changed here, I think that there is a change of line-endings or encoding. Could you check and potentially fix line-endings and encodings in this file?

File: domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs [hunk 19]
```diff
@@ -0,0 +1,15 @@
+using System;
+using Microsoft.EntityFrameworkCore;
+
+namespace API.Models;
+
+[Index(nameof(CompanyAId))]
+[Index(nameof(CompanyBId))]
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/Models/Connection.cs:7]: Here annotations are used, but in ApplicationDbContext, the fluent API is used for something similar. I suggest that you pick only one approach.

File: domains/transfer-agreements/TransferAgreements.API/API/swagger.io.apibeskrivelse.txt [hunk 20]
```diff
@@ -0,0 +1,184 @@
+openapi: 3.0.3
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API/swagger.io.apibeskrivelse.txt:1]: I know Christoffer wants this file here, but I think it should not be here. Or be in a different PR where the entire API is actually specified. 
But it is up to you to do what you want. You are welcome to ignore my comment here and click "resolve conversation".

File: domains/transfer-agreements/chart/API.migrations.sql [hunk 21]
```diff
@@ -0,0 +1,322 @@
+CREATE TABLE IF NOT EXISTS "__EFMigrationsHistory" (
```
Comment (duizer) [inline domains/transfer-agreements/chart/API.migrations.sql:1]: This file should be deleted

**Review by duizer — CHANGES_REQUESTED** [review]

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.Create_ShouldReturnInvitationId_WhenInvitationIsCreated.verified.txt [hunk 22]
```diff
@@ -0,0 +1 @@
+{"connectionInvitationId":"Guid_1"}
```
[the remark this answers is not in the document]
Comment (duizer) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.Create_ShouldReturnInvitationId_WhenInvitationIsCreated.verified.txt:1]: That must be in a different PR because the API endpoint does not return correct status code

File: domains/transfer-agreements/TransferAgreements.API/API/Migrations/ApplicationDbContextModelSnapshot.cs (hunk 18 above)
Comment (duizer) [reply domains/transfer-agreements/TransferAgreements.API/API/Migrations/ApplicationDbContextModelSnapshot.cs:1]: it is lf and utf-8 now, so that is good.

**Review by duizer — APPROVED** [review]

**Review by duizer — CHANGES_REQUESTED** [review]
Do stuff

**Review by duizer — DISMISSED** [review]

**Review by duizer — APPROVED** [review]
